File: reference/ts-server/src/server/rng.ts
export class Rng {
  private seed: number;

  public constructor(seed: number) {
    this.seed = seed >>> 0;
  }

  public next(): number {
    this.seed += 0x6d2b79f5;
    let t = this.seed;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  public int(min: number, max: number): number {
    return Math.floor(this.next() * (max - min + 1)) + min;
  }

  public pick<T>(values: T[]): T {
    return values[this.int(0, values.length - 1)] as T;
  }

  public bool(probability = 0.5): boolean {
    return this.next() < probability;
  }
}

File: rust/server/src/rng.rs
#[derive(Clone, Debug)]
pub struct Rng {
    seed: u32,
}

impl Rng {
    pub fn new(seed: u32) -> Self {
        Self { seed }
    }

    pub fn next_f32(&mut self) -> f32 {
        self.seed = self.seed.wrapping_add(0x6d2b79f5);
        let mut t = self.seed;
        t = (t ^ (t >> 15)).wrapping_mul(t | 1);
        t ^= t.wrapping_add((t ^ (t >> 7)).wrapping_mul(t | 61));
        let out = t ^ (t >> 14);
        (out as f64 / 4_294_967_296.0) as f32
    }

    pub fn int(&mut self, min: i32, max: i32) -> i32 {
        if max <= min {
            return min;
        }
        let span = (max - min + 1) as f32;
        min + (self.next_f32() * span).floor() as i32
    }

    pub fn bool(&mut self, probability: f32) -> bool {
        self.next_f32() < probability
    }

    pub fn pick_index(&mut self, len: usize) -> usize {
        if len <= 1 {
            return 0;
        }
        (self.next_f32() * len as f32).floor().min((len - 1) as f32) as usize
    }
}
